Add operational guide for autoscaling (#6801)
grafana/loki: docs/sources/operations/autoscaling_queriers.md

---
title: Autoscaling Loki Queriers
weight: 30
---

# Autoscaling Loki Queriers

A microservices deployment of a Loki cluster that runs on Kubernetes will typically handle a
workload that varies throughout the day.
To make Loki easier to operate and optimize the cost of running Loki at scale,
we have designed a set of resources to help you autoscale your Loki queriers.

## Prerequisites

You need to run Loki in Kubernetes as a set of microservices.

We recommend using [KEDA (*Kubernetes-based Event Driven Autoscaler*)](https://keda.sh/) to configure autoscaling
based on Prometheus metrics. Refer to the [KEDA documentation](https://keda.sh/docs/latest/deploy) to learn more
about setting up KEDA in your Kubernetes cluster.

> 

Reviewer: I think you forgot to mention in the prerequisites that users need to run the v2 of query frontend (as the metrics used for the scaling are only exposed from the query-scheduler)

Author: > the v2 of query frontend

@Raskyld Can you elaborate on what you mean by "v2" here?

@KMiller-Grafana should we add versioning requisites in these docs? IIUC, docs are coupled with the version (e.g. [docs for 2.4.x][1] and [docs for latest][2]).

[1]: https://grafana.com/docs/loki/v2.4.x/
[2]: https://grafana.com/docs/loki/latest/

Reviewer: Of course : there is two implementations of the `query-frontend` ([namely v1, v2](https://github.com/grafana/loki/tree/main/pkg/lokifrontend/frontend)) the first one has a builtin query queue and the second one use `scheduler_worker` (see [here](https://github.com/grafana/loki/blob/main/pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go)) to forward queries to the schedulers. 

*tl; dr* If you use the query schedulers you're using query-frontend v2 otherwise you fallback to the v1.

The v1 typically doesn't have the metric you're using for the auto-scaling, hence, it would be wise to add the query scheduler setup in the requirements.

If you need more information do not hesitate !

Author: > there is two implementations of the query-frontend ([namely v1, v2](https://github.com/grafana/loki/tree/main/pkg/lokifrontend/frontend))

TIL. I'll add a note to the prerequisites. Thank you! 🚀

Reviewer: You're very welcome! Thanks you :D

## Scaling metric

Queriers pull queries from the query scheduler queue and process them on the querier workers. Therefore, it makes sense to scale based on:

- The scheduler queue size.
- The queries running in the queriers.

The Query Scheduler exposes the `cortex_query_scheduler_inflight_requests` metric.
It tracks the number of queued queries plus the number of queries currently running in the querier workers.

> The following query is useful to scale queriers based on the inflight requests.

Author: > I'm suggesting that we use "query-scheduler," as most of the Mimir docs use a hyphenated form

That makes perfect sense.

>  Is it tracking the sum of the number of queued queries and the number of queries? Or is it tracking 2 distinct items?

It tracks the sum.

> "It tracks the sum of queued queries plus the number of queries currently running in the querier workers."

Sounds good to me!

> The following query is useful to scale queriers based on the inflight requests.

Author: Wrt to the suggested change (line 20)
```
Refer to the [Deploying KEDA] to learn more.
```

Should be 
```
Refer to the [Deploying KEDA section] to learn more.
```

wdyt?

Our recommendation is to adjust these values according to your workload.

## Cluster capacity planning

When configuring KEDA to scale your Loki queriers, we need to configure a few things:

- The threshold for scaling up and down.
- Scale down stabilization period.
- The minimum and the maximum number of queriers.

Querier workers process queries from the queue, each of our queriers can be configured to run several workers.
Our recommendation is to aim using 75% of those workers to leave some workforce headroom for workload spikes.
Therefore, if we configure our queriers to run 6 workers, we will set a threshold of `floor(0.75 * 6) = 4`.

Our recommendation for the minimum number of queriers is running at least one queriers and look at the average
number of inflight requests 75% of the time in the last seven days, targeting a 75% utilization of our queriers.
So if we use 6 workers per querier, we will use the following query:

```promql
clamp_min(ceil(
    avg(
        avg_over_time(cortex_query_scheduler_inflight_requests{namespace="loki-cluster", quantile="0.75"}[7d])
    ) / scalar(floor(vector(6 * 0.75)))
), 1)

> ```

Reviewer: If I understand correctly, you want that 75% of the time, the average number of request in the system (`L` in queuing theory) is equal to 75% the number of workers per querier, hence, there would be no query in the queue (they would be addressed immediately).
By evaluating the ratio between the **current** `L` (thanks to the inflight metric) and our **target** `L = 0.75 * nb_worker` you know how much times we need to scale the **current** number of querier to reach our **target**.

Side note, this is only meaningful if `lambda` (the request arrival rate) converge over time which is, I think, why you recommend using the average on a whole week.

So the result of this PromQL query isn't actually the minimum number of queriers necessary but the scale factor that we need to apply to the **current** number of queriers in the cluster when we evaluated the inflight metric.

**tl;dr** If I understood your methodology correctly, you still need to add `* nb_querier` for this query to give the minimal number of querier required for the cluster to instantly handle a query in 75% of the time.
**Or** you can remove the sentence "running **_at least one_** queriers" for "running only one querier" as it means `nb_querier` would be 1, but it's not practical as most users will already have a running cluster with a specific number of querier to maintain the cluster in an equilibrium state.

Please, let met know if I am wrong in my understanding of your approach and thanks a lot for your work, it saved me a bunch of time ! :rocket:

Author: > [...] you want that 75% of the time, the average number of request in the system (L in queuing theory) is equal to 75% the number of workers per querier

Not quite. Using the 75 percentile (75% of the time) is to avoid taking short-lasting spikes into account.

> If I understood your methodology correctly, you still need to add * nb_querier for this query

I don't think so. We divide by the number of workers we aim to use so we have extra room for spikes.
E.g. if the avg is 10 and we aim to use 2 workers out of the 4 workers available per querier, this query would return 5 (the number of queriers). Because we'd be using 50% of the workers (2 out of 4), we have room for twice the avg workload.

Reviewer: > @Raskyld Thanks for your comment! Sorry for the delay, I took some vacations.

No problem! Hope you enjoyed it :)

> Not quite. Using the 75 percentile (75% of the time) is to avoid taking short-lasting spikes into account.

Gotcha!

> I don't think so. We divide by the number of workers we aim to use so we have extra room for spikes. E.g. if the avg is 10 and we aim to use 2 workers out of the 4 workers available per querier, this query would return 5 (the number of queriers). Because we'd be using 50% of the workers (2 out of 4), we have room for twice the avg workload.

After two weeks and some sleep I realized my misunderstanding come from the metrics used.
Just for the record, and to be sure I understand correctly :

`cortex_query_scheduler_inflight_requests` observes (every `250ms`) the number of queries being processed by a querier worker or queued.
So you want the minimal number of worker = the average number of inflight requests 75% of the time ; then you divide the average by the number of active worker you want (e.g. 75% of all the workers) to know how much querier you need.

Am I correct? :pray: 

I have another concern : It is possible to replicate the `query-scheduler` as long as there is less or as much as instance of scheduler than instance of querier worker (i.e. `-querier.max-concurrent`) ; this means, you could have multiple series of this metric right? If so, shouldn't you sum the averages before dividing?

Author: > [...] the average number of inflight requests 75% of the time ; then you divide the average by the number of active worker you want (e.g. 75% of all the workers) to know how much querier you need.
> Am I correct? 

Yes! 

> you could have multiple series of this metric right? If so, shouldn't you sum the averages before dividing?

The `avg` is an aggregation function that will average all the series. To have different averages for each series you would do `avg by (instance)`.

There is an example [in the docs](https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators) that shows this:
> If the metric http_requests_total had time series that fan out by application, instance, and group labels, we could calculate the total number of seen HTTP requests per application and group over all instances via:
> `sum without (instance) (http_requests_total)`
> Which is equivalent to:
> `sum by (application, group) (http_requests_total)`
> If we are just interested in the total of HTTP requests we have seen in all applications, we could simply write:
> `sum(http_requests_total)`

Reviewer: Guess it's my turn to say TIL!
`avg_over_time` uses an aggregation operator but does **not** aggregate while `avg` alone does!

Reviewer: Thanks for all those explanations :ok_hand:

---
title: Autoscaling Loki queriers
menuTitle: Autoscaling queriers
description: Kubernetes deployments of a microservices mode Loki cluster can use KEDA to autoscale the quantity of queriers.
weight: 30
---

# Autoscaling Loki queriers

> A microservices deployment of a Loki cluster that runs on Kubernetes will typically handle a

Reviewer 2: ```suggestion
A microservices deployment of a Loki cluster that runs on Kubernetes typically handles a
```

workload that varies throughout the day.
To make Loki easier to operate and optimize the cost of running Loki at scale,
we have designed a set of resources to help you autoscale your Loki queriers.

## Prerequisites

You need to run Loki in Kubernetes as a set of microservices. You need to use the query-scheduler.

We recommend using [Kubernetes Event-Driven Autoscaling (KEDA)](https://keda.sh/) to configure autoscaling

> based on Prometheus metrics. Refer to the [Deploying KEDA](https://keda.sh/docs/latest/deploy) section to learn more

Reviewer 2: ```suggestion
based on Prometheus metrics. Refer to [Deploying KEDA](https://keda.sh/docs/latest/deploy) to learn more
```

about setting up KEDA in your Kubernetes cluster.

## Scaling metric

> Queriers pull queries from the query-scheduler queue and process them on the querier workers. Therefore, it makes sense to scale based on:

Reviewer 2: ```suggestion
Because queriers pull queries from the query-scheduler queue and process them on the querier workers, you should scale metrics based on:
```

- The scheduler queue size.
- The queries running in the queriers.

The query-scheduler exposes the `cortex_query_scheduler_inflight_requests` metric.
It tracks the sum of queued queries plus the number of queries currently running in the querier workers.
The following query is useful to scale queriers based on the inflight requests.

```promql
sum(
  max_over_time(
    cortex_query_scheduler_inflight_requests{namespace="loki-cluster", quantile="<Q>"}[<R>]
  )
)
```

> We can use the quantile (Q) and the range (R) parameters to fine-tune the metric.

Reviewer 2: ```suggestion
Use the quantile (Q) and the range (R) parameters to fine-tune the metric.
```

> We can use the quantile (Q) and the range (R) parameters to fine-tune the metric.

Reviewer 2: My suggestion below assumes that the user fine-tunes the metric. I'm not sure if my assumption is correct.

Author: That's correct.

The higher Q is, the more sensitive the metric is to short-lasting spikes.

> As R increases, we reduce the variation over time in the metric.

Reviewer 2: ```suggestion
As R increases, you can reduce the variation over time in the metric.
```

> As R increases, we reduce the variation over time in the metric.

Reviewer 2: I'm confused about 'we' in this sentence. Does the system automatically reduce variation?

Author: No. The user can modify Q and R to fine-tune the metric.

> A higher value for R helps avoid the autoscaler from modifying the number of replicas too frequently.

Reviewer 2: ```suggestion
A higher R-value helps avoid the autoscaler from modifying the number of replicas too frequently.
```

In our experience, we have found that a Q of 0.75 and an R of 2 minutes work well.

> Our recommendation is to adjust these values according to your workload.

Reviewer 2: ```suggestion
You can adjust these values according to your workload.
```

Author: "Our recommendation is to..." wording comes from feedback from Karen. I'm ok with reverting that, but just a note.

## Cluster capacity planning

> When configuring KEDA to scale your Loki queriers, we need to configure:

Reviewer 2: ```suggestion
To scale the Loki queries, you configure the following settings:
```

> When configuring KEDA to scale your Loki queriers, we need to configure:

Reviewer 2: Confused by we...is it truly a 'we' as in Grafana Labs is configuring KEDA? Or is this something the user does? I'll be able to add a suggestion after I get clarity.

Author: It's something the user does.

- The threshold for scaling up and down
- The scale down stabilization period
- The minimum and the maximum number of queriers

> Querier workers process queries from the queue. Each of our queriers can be configured to run several workers.

Reviewer 2: ```suggestion
Querier workers process queries from the queue. You can configure each Loki querier to run several workers.
```

> Our recommendation is to aim for using 75% of those workers, to leave some workforce headroom for workload spikes.

Reviewer 2: ```suggestion
To reserve workforce headroom to address workload spikes, do not use more than 75% of the workers. 
```

Author: I would rather write:
```
To reserve workforce headroom to address workload spikes, our recommendation is to not use more than 75% of the workers.
```

> Therefore, if we configure our queriers to run 6 workers, we will set a threshold of `floor(0.75 * 6) = 4`.

Reviewer 2: ```suggestion
For example, if you configure the Loki queriers to run 6 workers, set a threshold of `floor(0.75 * 6) = 4`.
```

> Our recommendation for the minimum number of queriers is to run at least one querier, and to look at the average
> number of inflight requests 75% of the time in the last seven days, targeting a 75% utilization of the queriers.

Reviewer 2: ```suggestion
To determine the minimum number of queries that you should run, run at least one querier for seven days and determine the average
number of inflight requests the system processes 75% of the time. The target utilization of the queries is 75%.
```

Author: > run at least one querier for seven days

This doesn't sound correct to me. Our recommendation is to run at least one querier (hence the `clamp_min`). but running one querier for seven days.

Reviewer 2: @salvacorts , updated the suggestion. Does that wording work better?

Author: Works perfectly! Thank you :)

> number of inflight requests 75% of the time in the last seven days, targeting a 75% utilization of the queriers.

Reviewer 2: I'm not sure what these lines mean. Is the user attempting to determine the minimum number of queriers to run?

Author: Correct. These are recommendations about how to estimate the minimum number of queriers to run.

So if we use 6 workers per querier, we will use the following query:

```promql
clamp_min(ceil(
    avg(
        avg_over_time(cortex_query_scheduler_inflight_requests{namespace="loki-cluster", quantile="0.75"}[7d])
    ) / scalar(floor(vector(6 * 0.75)))
), 1)
```

> Our recommendation for the maximum number of queriers is running the number of queriers that would be required

Reviewer 2: Similar question here, is the user trying to determine the max number of queriers?

Author: Yes

> to process all the inflight requests you had 50% of the time during the last seven days.

Reviewer 2: The maximum number of queriers to run is equal to the number of queriers required to process 50% of the inflight requests during a seven-day timespan.

Author: This is not correct: "50% of the inflight requests" is not the same as "all the inflight requests you had 50% of the time". The latter refers to a quantile; the 50 percentile is the median but not half.

Reviewer 2: Gotcha, ok what about "The maximum number of queriers to run is equal to the number of queriers required to process all inflight requests 50% of the time during a seven-day timespan."

does that work?

Author: That sounds perfect. Committed the changes. Thanks!

> As for the previous example, if each querier runs 6 workers, we divide the inflight requests by 6.

Reviewer 2: ```suggestion
As for the previous example, if each querier runs 6 workers, divide the inflight requests by 6.
```

The resulting query becomes:

```promql
ceil(
    max(
        max_over_time(cortex_query_scheduler_inflight_requests{namespace="loki-cluster", quantile="0.5"}[7d])
    ) / 6
)
```

> Finally, to minimize the scenario where we scale up shortly after scaling down, we recommend setting

Reviewer 2: ```suggestion
To minimize the scenario where Loki scales up shortly after scaling down, we recommend setting
```

a stabilization window for scaling down.

### KEDA configuration

This [KEDA ScaledObject](https://keda.sh/docs/latest/concepts/scaling-deployments/) example configures autoscaling
for the querier deployment in the `loki-cluster` namespace.

> We set the minimum number of replicas to 10 and the maximum number of replicas to 50.

Reviewer 2: ```suggestion
The example shows the minimum number of replicas set to 10 and the maximum number of replicas set to 50.
```

> Since each querier runs 6 workers, aiming to use 75% of those workers, we set the threshold to 4.

Reviewer 2: ```suggestion
Because each querier runs 6 workers, aiming to use 75% of those workers, the threshold is set to 4.
```

The metric is served at `http://prometheus.default:9090/prometheus`. We configure a stabilization window of 30 minutes.

```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: querier
  namespace: loki-cluster
spec:
  maxReplicaCount: 50
  minReplicaCount: 10
  scaleTargetRef:
    kind: Deployment
    name: querier
  triggers:
  - metadata:
      metricName: querier_autoscaling_metric
      query: sum(max_over_time(cortex_query_scheduler_inflight_requests{namespace="loki-cluster", quantile="0.75"}[2m]))
      serverAddress: http://prometheus.default:9090/prometheus
      threshold: "4"
    type: prometheus
  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 1800
```

## Prometheus alerting when at capacity

> Because the configured maximum may not be enough, a Prometheus alert can identify

Reviewer 2: ```suggestion
Because the configured maximum might not be sufficient, a Prometheus alert can identify
```

> when the quantity of queriers has been at its configured maximum for an extended time. This example specifies 3 hours as the extended time:

Reviewer 2: ```suggestion
when the quantity of queriers has been at its configured maximum for an extended time. The following example specifies three hours (`3h`) as the extended time:
```
